src/pybind/mgr/cli_api/module.py

```diff
@@ -1,92 +1,121 @@
+                        f'_cli_{name}',
+                        command)
+        return klass
+
+    @staticmethod
+    def is_public(func: Callable) -> bool:
+        return (
+            inspect.isfunction(func)
+            and not func.__name__.startswith('_')
+            and not API.hook.get(func)
+            and not API.internal.get(func)
+        )
+
+    @staticmethod
+    def func_wrapper(func):
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            return HandleCommandResult(
+                0,
+                json.dumps(func(self, *args, **kwargs), sort_keys=True, indent=4)
+            )
+
+        # functools doesn't change the signature when wrapping a function
+        # so we do it manually
+        signature = inspect.signature(func)
+        wrapper.__signature__ = signature
+        return wrapper
+
+
+class CLI(MgrModule, metaclass=MgrAPIReflector):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @CLICommand('mgr cli_benchmark')
+    def benchmark(self, iterations: int, threads: int, func_name: str, func_args: List[str] = None) -> HandleCommandResult:
+        if(iterations and threads):
+            try:
+                func = getattr(self, func_name)
+            except Exception:
+                return HandleCommandResult(stderr="Could not find the public function you are requesting")
         else:
-            raise BenchmarkException("Number of Total and number of parallel calls must be greater than 0")
+            raise BenchmarkException(
+                "Number of Total calls and number of parallel calls must be greater than 0")
+
+        def timer(i):
+            time_start = get_time()
+            if (func_args) is None:
+                func()
+            else:
+                func(*func_args)
+            return get_time() - time_start
 
-    def join(self):
-        for thread in self._threads:
-            thread.join()
+        with concurrent.futures.ThreadPoolExecutor(max_workers=threads) as executor:
+            results_iter = executor.map(timer, range(iterations))
+        results = list(results_iter)
 
-    def get_stats(self):
         stats = {
-            "avg": (self._time / self._number_of_total_calls),
-            "min": min(self._self_time),
-            "max": max(self._self_time)
+            "avg": sum(results) / len(results),
+            "max": max(results),
+            "min": min(results),
         }
-        return stats
-
-    def timer(self, jobs, func):
-        self._lock.acquire()
-        while not self._jobs.empty():
-            jobs.get()
-            time_start = time.time()
-            func()
-            time_end = time.time()
-            self._self_time.append(time_end - time_start)
-            self._time += (time_end - time_start)
-            self._jobs.task_done()
-        self._lock.release()
+        return HandleCommandResult(stdout=json.dumps(stats, indent=4))
```

Comment:
In other place you're using `sort_keys=True`, so to avoid mismatches it's better to encapsulate this:
```py
def pretty_json(obj):
    import json
    return json.dumps(obj, indent=4, sort_keys=True)
```